data-service/src/main/java/uk/gov/gchq/palisade/service/data/service/AuditableDataService.java

```diff
@@ -0,0 +1,96 @@
+ */
+public class AuditableDataService {
+    private final DataService dataService;
+
+    /**
+     * AuditableDataService constructor
+     *
+     * @param dataService the current Data Service implemetation
+     */
+    public AuditableDataService(final DataService dataService) {
+        this.dataService = dataService;
+    }
+
+    /**
+     * Provides a wrapped message with the reference to the resources that are to be provided to the client or an
+     * error message
+     *
+     * @param dataRequest request information from the client
+     * @return reference to the resource information or error message
+     */
+    public CompletableFuture<AuditableAuthorisedDataRequest> authoriseRequest(final DataRequest dataRequest) {
+        return dataService.authoriseRequest(dataRequest)
+                .thenApply(dataResponse -> AuditableAuthorisedDataRequest.Builder.create()
+                        .withDataRequest(dataRequest)
+                        .withDataResponse(dataResponse)
+                        .withErrorMessage(null))
+                .exceptionally(e -> AuditableAuthorisedDataRequest.Builder.create()
+                        .withDataRequest(dataRequest)
+                        .withDataResponse(null)
+                        .withErrorMessage(AuditErrorMessage.Builder.create(dataRequest)
+                                .withAttributes(Collections.singletonMap("method", "authoriseRequest"))
+                                .withError(e)));
+    }
+
+    /**
+     * Reads the authorised resource and passes this onto the client in the form an {@link OutputStream}.  The response
+     * is used in the construction of the audit message for this request.
+     *
+     * @param auditableAuthorisedDataRequest provides the reference to the authorised data request
+     * @param outputStream         is used to provide the requested data to be forwarded to the client
+     * @return information on the resources that have been provided
+     */
+    public CompletableFuture<AuditableDataResponse> read(final AuditableAuthorisedDataRequest auditableAuthorisedDataRequest, final OutputStream outputStream) {
+        DataRequest dataRequest = auditableAuthorisedDataRequest.getDataRequest();
+        AuthorisedData authorisedData = auditableAuthorisedDataRequest.getAuthorisedData();
+        AtomicLong recordsProcessed = new AtomicLong(0);
+        AtomicLong recordsReturned = new AtomicLong(0);
+
+        return dataService.read(authorisedData, outputStream, recordsProcessed, recordsReturned)
+                .thenApply(pair -> AuditableDataResponse.Builder.create()
+                        .withToken(dataRequest.getToken())
+                        .withSuccessMessage(AuditSuccessMessage.Builder.create(auditableAuthorisedDataRequest)
+                                .withRecordsProcessedAndReturned(recordsProcessed.get(), recordsReturned.get()))
+                        .withAuditErrorMessage(null))
```

Comment:
The auditable types will hold data depending on the outcome of a process.  In this method, the AuditableDataResponse will have have a AuditSuccessMessage and if during the process it fails, it will have an AuditErrorMessage.  So for the builder I want the builder to work either way

Reply:
Would be better to have `withNoAuditErrorMesaage()` if you really want to call a method when there is none.